### ISSUE TITLE
Implementing an adjoint calculation for backprop-ing through time

### ISSUE THREAD

**Issue by ianwilliamson** [fancompute/wavetorch#1]
Should consider the performance benefit of implementing an adjoint calculation for the backward pass through the `forward()` method in `WaveCell`. This would potentially save us on memory during gradient computation because pytorch doesn't need to construct as large of a graph. 

The approach is described here: https://pytorch.org/docs/stable/notes/extending.html

**Comment by parenthetical-e** [comment]
Sorry to pop in, but on the off and maybe small chance you folks haven’t seen this lib/paper:

https://github.com/rtqichen/torchdiffeq
https://arxiv.org/pdf/1806.07366.pdf

Implements a ODE solver and uses adjoint methods for the backward pass. This is what you need? 

I was already thinking about porting WaveCell to it for my own use. Collaborate?

**Comment by twhughes** [comment]
Hey Eric could you forward that email to me as well please?  Im interested
in what you have planned. Thanks!

**Comment by ianwilliamson** [comment]
This is now partially implemented. Currently, the individual time step is a primitive. This seems to help with memory utilization during training, especially with nonlinearity. Perhaps we could investigate if there would be significant performance benefits from adjoint-ing the time loop as well.
